File: festival-web/src/main/java/cn/bdqfork/web/route/RouteHandler.java
package cn.bdqfork.web.route;

import cn.bdqfork.core.util.ReflectUtils;
import cn.bdqfork.web.constant.ContentType;
import cn.bdqfork.web.route.message.HttpMessageHandler;
import cn.bdqfork.web.route.response.ResponseHandlerFactory;
import io.vertx.core.Handler;
import io.vertx.ext.web.RoutingContext;

import java.lang.reflect.Method;

/**
 * @author bdq
 * @since 2020/2/21
 */
public class RouteHandler implements Handler<RoutingContext> {
    private HttpMessageHandler httpMessageHandler;
    private ResponseHandlerFactory responseHandlerFactory;
    private Method method;
    private Object bean;

    public RouteHandler(HttpMessageHandler httpMessageHandler, ResponseHandlerFactory responseHandlerFactory, Method method, Object bean) {
        this.httpMessageHandler = httpMessageHandler;
        this.responseHandlerFactory = responseHandlerFactory;
        this.method = method;
        this.bean = bean;
    }

    @Override
    public void handle(RoutingContext routingContext) {
        try {
            Object[] args = httpMessageHandler.handle(routingContext, method.getParameters());
            Object result = ReflectUtils.invokeMethod(bean, method, args);
            if (ReflectUtils.isReturnVoid(method)) {
                return;
            }
            String contentType = routingContext.getAcceptableContentType();
            if (result instanceof ModelAndView) {
                contentType = ContentType.HTML;
            }
            responseHandlerFactory.getResponseHandler(contentType).handle(routingContext, result);
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

}

File: festival-kotlin-web/src/main/kotlin/cn/bdqfork/kotlin/web/route/RouteHandler.kt
package cn.bdqfork.kotlin.web.route

import cn.bdqfork.core.util.ReflectUtils
import cn.bdqfork.kotlin.web.constant.ContentType
import cn.bdqfork.kotlin.web.route.message.HttpMessageHandler
import cn.bdqfork.kotlin.web.route.response.ResponseHandlerFactory
import io.vertx.core.Handler
import io.vertx.ext.web.RoutingContext
import java.lang.reflect.Method

/**
 * @author bdq
 * @since 2020/2/18
 */
class RouteHandler(private val httpMessageHandler: HttpMessageHandler, private val responseHandlerFactory: ResponseHandlerFactory, private val method: Method, private val bean: Any) : Handler<RoutingContext> {
    override fun handle(routingContext: RoutingContext) {
        try {
            val args = httpMessageHandler.handle(routingContext, method.parameters)
            val result = ReflectUtils.invokeMethod(bean, method, *args)
            if (ReflectUtils.isReturnVoid(method)) {
                return
            }
            var contentType = routingContext.acceptableContentType
            if (result is ModelAndView) {
                contentType = ContentType.HTML
            }
            responseHandlerFactory.getResponseHandler(contentType).handle(routingContext, result)
        } catch (e: Exception) {
            throw IllegalStateException(e)
        }
    }

}